arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -637,64 +835,161 @@ void RocksDBMetaCollection::corruptRevisionTree(uint64_t count, uint64_t hash) {
+    auto lockGuard = scopeGuard([this] { unlockWrite(); });
+    // get the exclusive lock on the collection, so that no new write transactions
+    // can start for this collection
+    auto res = lockWrite(transaction::Options::defaultLockTimeout);
+    if (res != TRI_ERROR_NO_ERROR) {
+      lockGuard.cancel();
+      THROW_ARANGO_EXCEPTION(res);
     }
-
-    _revisionTree = std::make_unique<RevisionTreeAccessor>(std::move(newTree), _logicalCollection);
-    _revisionTreeApplied = beginSeq;
-    _revisionTreeCreationSeq = beginSeq;
-    _revisionTreeSerializedSeq = beginSeq;
-
-    {
-      // finally remove all pending updates up to including our own sequence number
-      std::unique_lock<std::mutex> guard(_revisionBufferLock);
     
+    // we now have the exclusive lock on the collection!
+    
+    // utility function to remove buffered updates up to (including) seq.
+    // requires the buffer lock to be held!
+    auto removeBufferedUpdatesUpTo = [this](rocksdb::SequenceNumber seq) {
       {
         auto it = _revisionTruncateBuffer.begin(); 
-        while (it != _revisionTruncateBuffer.end() && *it <= beginSeq) {
+        while (it != _revisionTruncateBuffer.end() && *it <= seq) {
           it = _revisionTruncateBuffer.erase(it);
         }
       }
-        
+      
       {
         auto it = _revisionInsertBuffers.begin(); 
-        while (it != _revisionInsertBuffers.end() && it->first <= beginSeq) {
+        while (it != _revisionInsertBuffers.end() && it->first <= seq) {
           it = _revisionInsertBuffers.erase(it);
         }
       }
-        
+      
       {
         auto it = _revisionRemovalBuffers.begin(); 
-        while (it != _revisionRemovalBuffers.end() && it->first <= beginSeq) {
+        while (it != _revisionRemovalBuffers.end() && it->first <= seq) {
           it = _revisionRemovalBuffers.erase(it);
         }
       }
+    };
+    
+    // place a blocker and add a guard to remove it at the end
+    TransactionId trxId = TransactionId(transaction::Context::makeTransactionId());
+  
+    auto blockerGuard = scopeGuard([&] {  // remove blocker afterwards
+      removeRevisionTreeBlocker(trxId);
+    });
+      
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
+    }
+    
+    // we now have a blocker in place that will prevent updates from being 
+    // applied
+    rocksdb::SequenceNumber blockerSeq = placeRevisionTreeBlocker(trxId);
+    
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
     }
 
-    return {};
+    {
+      // we may still have some buffered updates, so let's remove them.
+      std::unique_lock<std::mutex> guard(_revisionTreeLock);
+      removeBufferedUpdatesUpTo(blockerSeq - 1);
```

Comment:
That is actually fine. We only get here in case we have a corrupted tree, and we will rebuild it from the collection data using a new snapshot. So removing everything before the snapshot is ok.